src/tray/mod.rs

```diff
@@ -0,0 +1,424 @@
+        }
+
+        while let Ok(msg) = receiver.recv() {
+            match msg {
+                TrayMessage::Enable => unsafe {
+                    if window.is_none() {
+                        create_window(true);
+                    }
+                },
+                TrayMessage::Disable => unsafe {
+                    if let Some(window) = window.take() {
+                        PostMessageW(window.0, WM_CLOSE, 1, 0);
+                    }
+                },
+                TrayMessage::CloseToTray => unsafe {
+                    if let Some(window) = window.as_ref() {
+                        PostMessageW(window.0, WM_HIDE_GUI, 0, 0);
+                    }
+                },
+                TrayMessage::TrayCreated(win) => {
+                    window = Some(win);
+                }
+            }
+        }
+    });
+}
+
+unsafe fn create_window(show_balloon: bool) {
+    thread::spawn(move || {
+        let mut tray_state = TrayState {
+            gui_handle: None,
+            gui_hidden: false,
+            about_shown: false,
+            close_gui: false,
+            show_balloon,
+        };
+
+        // Keep searching for window handle until its found
+        while tray_state.gui_handle.is_none() {
+            EnumWindows(Some(enum_proc), &mut tray_state as *mut _ as LPARAM);
+        }
+
+        let h_instance = GetModuleHandleW(ptr::null());
+
+        let class_name = str_to_wide!("Ajour Tray");
+
+        let mut class = mem::zeroed::<WNDCLASSEXW>();
+        class.cbSize = mem::size_of::<WNDCLASSEXW>() as u32;
+        class.lpfnWndProc = Some(window_proc);
+        class.hInstance = h_instance;
+        class.lpszClassName = class_name.as_ptr();
+        class.hbrBackground = CreateSolidBrush(RGB(0, 77, 128));
+
+        RegisterClassExW(&class);
+
+        let hwnd = CreateWindowExW(
+            WS_EX_NOACTIVATE,
+            class_name.as_ptr(),
+            ptr::null(),
+            0,
+            0,
+            0,
+            0,
+            0,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            h_instance,
+            &mut tray_state as *mut _ as *mut std::ffi::c_void,
+        );
+
+        // Send window handle back to main loop so we can post messages to it
+        let _ = TRAY_SENDER
+            .get()
+            .unwrap()
+            .try_send(TrayMessage::TrayCreated(WindowHandle(hwnd)));
+
+        let mut msg = mem::zeroed();
+        while GetMessageW(&mut msg, ptr::null_mut(), 0, 0) != 0 {
+            TranslateMessage(&msg);
+            DispatchMessageW(&msg);
+        }
+
+        if tray_state.close_gui {
+            // WM_QUIT was sent, which was triggered by the Exit button. Close entire program.
+            SHOULD_EXIT.store(true, Ordering::Relaxed);
+
+            // Activate window to force event loop to run / see that it should now close
+            ShowWindow(tray_state.gui_handle.unwrap(), SW_SHOWMINIMIZED);
```

Comment:
Can this `.unwrap()` fail

Reply:
Nope, it's safe. Lines 117-119 is where we find the gui handle and assign it to the state before creating our window. So within the windows loop, it'll always be Some otherwise we would have never made it to the point of making the window.

Reply:
See the `enum_proc` procedure and the https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumwindows documentation.

Since Iced doesn't give us a way to reference the underlying window handle (HWND), I have to use this function to find it. Basically it enumerates over all windows and I just find the "active" window that has the same process id as Ajour, lol.

Reply:
We need that window handle so we can show / hide it w/

```rust
let cmd = if state.gui_hidden {
                    SW_RESTORE
                } else {
                    SW_HIDE
                };

                ShowWindow(*state.gui_handle.as_ref().unwrap(), cmd)
```